rfd/00xx-snowflake-support.md

```diff
@@ -0,0 +1,171 @@
+
+#### User authentication
+
+Username can be extracted from `/session/v1/login-request` requests, field `data.LOGIN_NAME`. The extracted username
+can be then use to authenticate the user against the Teleport RBAC.
+
+### CLI support 
+
+`snowsql` is the official Snowflake CLI tool, and it doesn't support TLS terminated proxies as described 
+in https://docs.snowflake.com/en/user-guide/snowsql-start.html#using-a-proxy-server, and it uses plain text communication 
+when connecting to the Snowflake on a different port that 443. 
+
+`snowsql` requires to use one of the available authentication method. By default, it asks for password unless other 
+authentication option is specified. The other alternatives are SSO and JWT/Private key (https://docs.snowflake.com/en/user-guide/snowsql-start.html).
+Teleport could use JWT/Private key for authentication. In this flow `tsh` would export user's private in PKCS8 format (at that's the only one supported by `snowsql`),
+and `snowsql` could use it to sign the JWT token.
+
+I think there are two ways on how we can support the `snowsql`: 
+
+1. `snowcli` can connect to a custom host **without** TLS on any port different from 443. 
+   Because we cannot start `tsh proxy db` on port 443 (all systems require the root account for that) we could 
+   start `tsh` proxy on a different port as `tsh` can encrypt the transmission to the Teleport proxy. This approach of course
+   has one big drawback; the connection is not end-to-end encrypted as communication between `tsh` and `snowsql` uses plain text.
+2. Despite the documentation I was able to use TLS terminated proxy with `tsh`. In this flow `snowsql` sends the traffic 
+   to Teleport by using HTTP proxy protocol. Teleport then was able to terminate the TLS connection and read the traffic.
+   In order to make it to work I exported the Teleport issued CA as a `CURL_CA_BUNDLE` environment variable (this is undocumented feature).
+
+Below example shows how to start `snowsql` in that mode:
+
+```shell
+export https_proxy=localhost:2000
+export http_proxy=localhost:2000
+export CURL_CA_BUNDLE=/opt/homebrew/etc/ca-certificates/cert.pem # includes the Teleport issued CA
+export REQUESTS_CA_BUNDLE=/opt/homebrew/etc/ca-certificates/cert.pem
+
+snowsql --noup -a im56867.us-east-2.aws -u teleport-user --private-key-path ${HOME}/.tsh/keys/example.com/bob.pkcs8 -h me.localhost
+```
+
+* environment variable `https_proxy` - sets the address of `tsh proxy db` tunnel address. Traffic will be sent there.
+* environment variable `CURL_CA_BUNDLE` - passes Teleport's CA bundle, so we can terminate TLS connection with our certificate.
+* `--private-key-path` - sets the path to PCKS8 private key that will be used to sign JWT.
+* `-h me.localhost` - is the name of destination server, in our case this name has to match the name on the certificate that we will use for TLS termination.
+
+### UX
+
+Snowflake will integrate with teleport in the same way as other databases. 
```

Comment:
In that case where the Snowflake JWT token will be stored ? In app access we have backend App/WebSession but I don't think that we have  an equivalent of this backend type for database access. 
Generating JWT token on each request can drain to Snowflake quota where storing token in memory at first glance seems to be find but not sure a Snowflake token can generate and use N simultaneous tokens (HA case with meny Snowflake agents)

Reply:
https://github.com/gravitational/teleport/pull/12006#discussion_r858070934

Reply:
One thing that can be not obvious. All clients that I've tested so far keep connection open, so we can just keep the tokens in the memory until the connection ends.